python/ray/tune/search/vizier/vizier_search.py

```diff
@@ -0,0 +1,248 @@
+            )
+        trial_client = self._active_trials[trial_id]
+        elapsed_secs = (
+                datetime.datetime.now().astimezone()
+                - trial_client.materialize().creation_time
+        )
+        trial_client.add_measurement(
+            svz.Measurement(
+                {k: v for k, v in result.items() if isinstance(v, float)},
+                elapsed_secs=elapsed_secs.total_seconds(),
+            )
+        )
+
+    def on_trial_complete(
+        self, trial_id: str, result: Optional[Dict] = None, error: bool = False
+    ) -> None:
+        if trial_id not in self._active_trials:
+            raise RuntimeError(f'No active trial for {trial_id}')
+        if self.study_client is None:
+            raise RuntimeError(
+                'VizierSearch not initialized! Set a search space first.'
+            )
+        trial_client = self._active_trials[trial_id]
+
+        if error:
+            # Mark the trial as infeasible.
+            trial_client.complete(
+                infeasible_reason=f'Trial {trial_id} failed: {result}'
+            )
+        else:
+            measurement = None
+            if result:
+                elapsed_secs = (
+                        datetime.datetime.now().astimezone()
+                        - trial_client.materialize().creation_time
+                )
+                measurement = svz.Measurement(
+                    {k: v for k, v in result.items() if isinstance(v, float)},
+                    elapsed_secs=elapsed_secs.total_seconds(),
+                )
+            trial_client.complete(measurement=measurement)
+        self._active_trials.pop(trial_id)
+
+    def suggest(self, trial_id: str) -> Optional[Dict]:
+        suggestions = self._study_client.suggest(count=1, client_id=trial_id)
+        if not suggestions:
+            return Searcher.FINISHED
+
+        self._active_trials[trial_id] = suggestions[0]
+        return self._active_trials[trial_id].parameters
+
+    def save(self, checkpoint_path: str) -> None:
+        # We assume that the Vizier service continues running, so the only
+        # information needed to restore this searcher is the mapping from the Ray
+        # to Vizier trial ids. All other information can become stale and is best
+        # restored from the Vizier service in restore().
+        ray_to_vizier_trial_ids = {}
+        for trial_id, trial_client in self._active_trials.items():
+            ray_to_vizier_trial_ids[trial_id] = trial_client.id
+        with open(checkpoint_path, 'w') as f:
+            info = {'study_id': self._study_id, 'ray_to_vizier_trial_ids': ray_to_vizier_trial_ids}
+            json.dump(info, f)
+
+    def restore(self, checkpoint_path: str) -> None:
+        with open(checkpoint_path, 'r') as f:
+            obj = json.load(f)
+
+        self._study_id = obj['study_id']
+        self._study_client = clients.Study.from_owner_and_id('raytune', self._study_id)
```

Comment:
Is there any handling logic that needs to be done here if the client is not recoverable?

Reply:
Good ask - It'll give out a `KeyError` if the study never existed.

I sent an additional tidying PR from our side: https://github.com/google/vizier/pull/1200 to make this error instead our custom `ResourceNotFoundError`, but in any case, an Exception will be raised here. Is this fine? I didn't find any explicit error handling in the other searchers (e.g. ZooptSearch: https://github.com/ray-project/ray/blob/master/python/ray/tune/search/zoopt/zoopt_search.py)

Reply:
I think it's okay for now, maybe we can add a TODO here just to make it clearer in case this pops up as an issue in the future.

Reply:
Done, added TODO.